docs/language/informal/local-inference.md

```diff
@@ -0,0 +1,498 @@
+invocation of the method to find a (possibly partial) solution for the missing
+type arguments
+
+`[T0 -> P0, ..., Tn -> Pn]` is a partial solution for a set of type variables
+`<T0 extends B0, ..., Tn extends Bn>` under constraint set `Cp` given a type
+expectation of `R` with respect to a return type `Q` (in which the `Ti` may be
+free) where the `Pi` are type schemas (potentially just `?` if unresolved)/
+
+If `R <: Q [T0, ..., Tn] -> C` does not hold, then each `Pi` is `?` and `Cp` is
+    empty
+
+Otherwise:
+  - `R <: Q [T0, ..., Tn] -> C` and `Cp` is `C`
+  - If `C` does not constrain `Ti` then `Pi` is `?`
+  - If `C` partially constrains `Ti`
+    - If `C` is over constrained, then it is an inference failure error
+    - Otherwise `Pi` is the constraint solution for `Ti` with respect to `C`
+  - If `C` fully constrains `Ti`, then
+    - Let `Ai` be `Bi[R0/T0, ..., ?/Ti, ..., ?/Tn]`
+    - If `C + Ti <: Ai` is over constrained, it is an inference failure error.
+    - Otherwise, `Ti` is fixed to be `Pi`, where `Pi` is the constraint solution
+      for `Ti` with respect to `C + Ti <: Ai`.
+
+### Upwards generic instantiation resolution
+
+Upwards resolution is the process by which the parameter types of a generic
+method (or constructor, etc) are matched against the actual argument types from
+an invocation of a method to find a solution for the missing type arguments that
+have not been fixed by downwards resolution.
+
+`[T0 -> M0, ..., Tn -> Mn]` is the upwards solution for an invocation of a
+generic method of type `<T0 extends B0, ..., Tn extends Bn>(P0, ..., Pk) -> Q`
+given actual argument types `R0, ..., Rk`, a partial solution `[T0 -> P0, ...,
+Tn -> Pn]` and a partial constraint set `Cp`:
+  - If `Ri <: Pi [T0, ..., Tn] -> Ci`
+  - And the full constraint resolution of `Cp + C0 + ... + Cn` for `<T0 extends
+B0, ..., Tn extends Bn>` given the initial partial solution `[T0 -> P0, ..., Tn
+-> Pn]` is `[T0 -> M0, ..., Tn -> Mn]`
+
+### Discussion
+
+The incorporation of the type bounds information is asymmetric and procedural:
+it iterates through the bounds in order (`Bi[R0/T0, ..., ?/Ti, ..., ?/Tn]`).  Is
+there a better formulation of this that is symmetric but still allows some
+propagation?
+
+### Inference rules
+
+- The expression `e as T` is inferred as `m as T` of type `T` in context `K`:
+  - If `e` is inferred as `m` in an empty context
+- The expression `x = e` is inferred as `x = m` of type `T` in context `K`:
+  - If `e` is inferred as `m` of type `T` in context `M` where `x` has type `M`.
+- The expression `x ??= e` is inferred as `x ??= m` of type `UP(T, M)` in
+  context `K`:
+  - If `e` is inferred as `m` of type `T` in context `M` where `x` has type `M`.
+- The expression `await e` is inferred as `await m` of type `T` in context `K`:
+  - If `e` is inferred as `m` of type `T` in context `J` where:
+    - `J` is `FutureOr<K>` if `K` is not `_`, and is `_` otherwise
```

Comment:
FYI, analyzer's rule is:
- If `K` is `_`, `dynamic`, or `FutureOr<L>` for some `L`, `J` is `K`.
- Otherwise `J` is `FutureOr<K>`